### PR TITLE
Optimization parallel support

### PR DESCRIPTION
**Pull request by tholden**
Fixes https://github.com/DynareTeam/dynare/issues/1400
Also provides parallel support for CMAES.

See also my related pull requests to dates and dseries:
 * https://github.com/DynareTeam/dseries/pull/26
 * https://github.com/DynareTeam/dates/pull/9

### CODE REVIEW

File: matlab/optimization/parallel_wrapper.m [hunk 1]
```diff
@@ -0,0 +1,36 @@
+function RV = parallel_wrapper( objective_function, XV, varargin )
+    n = size( XV, 2 );
+    RV = zeros( 1, n );
+    
+    global objective_function_penalty_base
```
Comment (JohannesPfeifer) [inline matlab/optimization/parallel_wrapper.m]: @tholden We removed this global variable in  #1051. Why are you reintroducing it?

Comment (tholden) [reply matlab/optimization/parallel_wrapper.m]: Because I hadn't noticed/had forgotten that it had been removed! Fixed in the commit below.

**Comment by JohannesPfeifer** [comment]
@tholden We are very sorry this got lost. We were too busy getting 4.5.2 out. The main issue was that that there had been a massive change to `dates` and `dseries` outstanding for the unstable branch that has now been merged. This is what creates the merge conflict. Do you by any chance know whether these changes resolved the required changes for the submodules?

Apart from this, I wanted to have a look at this interesting PR for quite some time, but teaching commitments intervened. The issue I am having is that you did not document anything here, Thus, I have a hard time following what you are actually doing and how I could test this. In particular, I do not know how to invoke a parallel mode-finding.

**Comment by tholden** [comment]
You've switched from the old-oop-style branch? The master branch requires different fixes I seem to remember. I'm afraid I may not be able to update this till December as I've got two papers to finish by the end of the month for a funder.
It is set up so parallel CMAES is invoked using 'UseParallel', just as it is for any other solver. See: https://github.com/DynareTeam/dynare/pull/1508/commits/bdfb68fa7fb488f254f2975beac1920b6eea455e
But yes, I should have explained this in the original PR.

**Comment by ivanlozev** [comment]
I would like ask what has happened to this branch, a sit is closed now? Will it be merged in the next Dynare version or it is discontinued? I am asking, because I want to implement simple GMM through fmincon, but I am having trouble running fmincon through MultiStart with the option ‘UseParallel’, true. This problem seems to be connected to the elements of class ‘dates’ being converted to structures.  Is there a workaround for that or I better wait for a fix in Dynare?

**Comment by tholden** [comment]
I have a stable Dynare release with this feature available from here: https://github.com/tholden/dynare/releases

**Pull request closed without merging**